decimal/bcd.go

```diff
@@ -0,0 +1,255 @@
+// |        0xd6       |    0x01    | 0x02  | 0x01 | 0x23 | 0x4d       |
+//
+// The decimal 0.000000000000000000000000000000000010 will be encoded as
+// 0xc7, 0x03, 0x01, 0x24, 0x01, 0x0c:
+//
+// | MP_EXT (ext 8) | length | MP_DECIMAL | scale |  1   | 0 (plus) |
+// |      0xc7      |  0x03  |    0x01    | 0x24  | 0x01 | 0x0c     |
+//
+// See also:
+//
+// * MessagePack extensions https://www.tarantool.io/en/doc/latest/dev_guide/internals/msgpack_extensions/
+//
+// * An implementation in C language https://github.com/tarantool/decNumber/blob/master/decPacked.c
+package decimal
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	bytePlus  = byte(0x0c)
+	byteMinus = byte(0x0d)
+)
+
+var isNegative = [256]bool{
+	0x0a: false,
+	0x0b: true,
+	0x0c: false,
+	0x0d: true,
+	0x0e: false,
+	0x0f: false,
+}
+
+// Calculate a number of digits in a buffer with decimal number.
+//
+// Plus, minus, point and leading zeroes do not count.
+// Contains a quirk for a zero - returns 1.
+//
+// Examples (see more examples in tests):
+//
+// - 0.0000000000000001 - 1 digit
+//
+// - 00012.34           - 4 digits
+//
+// - 0.340              - 3 digits
+//
+// - 0                  - 1 digit
+func GetNumberLength(buf string) int {
+	if len(buf) == 0 {
+		return 0
+	}
+	n := 0
+	for _, ch := range []byte(buf) {
+		if ch == '0' && n != 0 {
+			n += 1
+			continue
+		}
+		if ch >= '1' && ch <= '9' {
+			n += 1
+		}
+	}
+
+	// Fix a case with a single 0.
+	if n == 0 {
+		n = 1
+	}
+
+	return n
+}
+
+// EncodeStringToBCD converts a string buffer to BCD Packed Decimal.
+//
+// The number is converted to a BCD packed decimal byte array, right aligned in
+// the BCD array, whose length is indicated by the second parameter. The final
+// 4-bit nibble in the array will be a sign nibble, 0x0c for "+" and 0x0d for
+// "-". Unused bytes and nibbles to the left of the number are set to 0. scale
+// is set to the scale of the number (this is the exponent, negated).
+func EncodeStringToBCD(buf string) ([]byte, error) {
+	if len(buf) == 0 {
+		return nil, fmt.Errorf("Length of number is zero")
+	}
+	signByte := bytePlus // By default number is positive.
+	if buf[0] == '-' {
+		signByte = byteMinus
+	}
+
+	// The first nibble should contain 0, if the decimal number has an even
+	// number of digits. Therefore highNibble is false when decimal number
+	// is even.
+	highNibble := true
+	l := GetNumberLength(buf)
+	if l%2 == 0 {
+		highNibble = false
+	}
+	scale := 0 // By default decimal number is integer.
+	var byteBuf []byte
+	for i, ch := range []byte(buf) {
+		// Skip leading zeroes.
+		if (len(byteBuf) == 0) && ch == '0' {
+			continue
+		}
+		if (i == 0) && (ch == '-' || ch == '+') {
+			continue
+		}
+		if ch == '.' && scale != 0 {
+			return nil, fmt.Errorf("Number contains more than one point")
+		}
+		// Calculate a number of digits after the decimal point.
+		if ch == '.' {
+			scale = len(buf) - i - 1
+			continue
+		}
```

Comment:
Not critical and just a matter of taste [2]
```suggestion
		// Calculate a number of digits after the decimal point.
		if ch == '.' {
			if scale != 0 {
				return nil, fmt.Errorf("Number contains more than one point")
			}
			scale = len(buf) - i - 1
			continue
		}
```

Reply:
I don't see a big difference, but changed it for you.